File: backend/src/aithru_agent/runtime/processors/usage.py
from __future__ import annotations

from aithru_agent.domain import (
    AgentRun,
    AgentRunBudgetStatus,
    AgentRunTreeUsageSnapshot,
    AgentRunUsageSummary,
)
from aithru_agent.domain.usage import aggregate_model_usage_payloads, evaluate_budget_status
from aithru_agent.persistence.protocols import AgentEventStore, AgentStore
from aithru_agent.stream.events import AgentStreamEvent


LINEAGE_CHILD_EVENT_TYPES = {
    "operator_action.follow_up.created",
    "research.continuation.created",
}


async def build_run_usage_summary(
    run: AgentRun,
    event_store: AgentEventStore,
) -> AgentRunUsageSummary:
    events = await event_store.list_by_run(run.id)
    return aggregate_model_usage_payloads(
        run.id,
        [
            event.payload
            for event in events
            if event.type == "model.usage" and isinstance(event.payload, dict)
        ],
        budget_policy=run.harness_options.budget_policy if run.harness_options else None,
        model_cost_policy=(
            run.harness_options.model_cost_policy if run.harness_options else None
        ),
    )


async def build_run_tree_usage_snapshot(
    root_run: AgentRun,
    store: AgentStore,
    event_store: AgentEventStore,
) -> AgentRunTreeUsageSnapshot:
    runs_by_id = {run.id: run for run in await store.list_runs()}
    runs_by_id[root_run.id] = root_run
    ordered_run_ids: list[str] = []
    summaries_by_id: dict[str, AgentRunUsageSummary] = {}
    visiting: set[str] = set()

    async def visit(run: AgentRun) -> AgentRunUsageSummary:
        if run.id in summaries_by_id:
            return summaries_by_id[run.id]
        if run.id in visiting:
            return await build_run_usage_summary(run, event_store)
        visiting.add(run.id)
        ordered_run_ids.append(run.id)
        direct = await build_run_usage_summary(run, event_store)
        child_summaries: list[AgentRunUsageSummary] = []
        subagents = await store.list_subagent_runs(parent_run_id=run.id)
        child_run_ids = [
            subagent.child_run_id
            for subagent in sorted(subagents, key=lambda item: item.id)
        ]
        child_run_ids.extend(
            _lineage_child_run_ids(await event_store.list_by_run(run.id))
        )
        seen_child_run_ids: set[str] = set()
        for child_run_id in child_run_ids:
            if child_run_id in seen_child_run_ids:
                continue
            seen_child_run_ids.add(child_run_id)
            if child_run_id in visiting or child_run_id in summaries_by_id:
                continue
            child = runs_by_id.get(child_run_id) or await store.get_run(child_run_id)
            if child is None:
                continue
            runs_by_id[child.id] = child
            child_summaries.append(await visit(child))
        summary = _with_descendant_usage(direct, child_summaries)
        summaries_by_id[run.id] = summary
        visiting.remove(run.id)
        return summary

    root_summary = await visit(root_run)
    summaries = [summaries_by_id[run_id] for run_id in ordered_run_ids if run_id in summaries_by_id]
    return AgentRunTreeUsageSnapshot(
        root_run_id=root_run.id,
        runs=summaries,
        total_requests=root_summary.total_requests,
        total_tokens=root_summary.total_tokens,
        total_model_cost_usd=root_summary.total_model_cost_usd,
        budget_status=_worst_budget_status(summaries),
        warnings=sorted({warning for summary in summaries for warning in summary.warnings}),
    )


def _with_descendant_usage(
    summary: AgentRunUsageSummary,
    child_summaries: list[AgentRunUsageSummary],
) -> AgentRunUsageSummary:
    descendant_requests = sum(child.total_requests for child in child_summaries)
    descendant_total_tokens = sum(child.total_tokens for child in child_summaries)
    descendant_model_cost_usd = sum(child.total_model_cost_usd for child in child_summaries)
    budget_status, warnings = evaluate_budget_status(
        requests=summary.own_requests + descendant_requests + summary.external_requests,
        total_tokens=summary.own_total_tokens
        + descendant_total_tokens
        + summary.external_total_tokens,
        model_cost_usd=summary.own_model_cost_usd
        + descendant_model_cost_usd
        + summary.external_model_cost_usd,
        budget_policy=summary.budget_policy,
        model_cost_policy=summary.model_cost_policy,
    )
    return summary.model_copy(
        update={
            "descendant_requests": descendant_requests,
            "descendant_input_tokens": sum(
                child.own_input_tokens + child.descendant_input_tokens
                for child in child_summaries
            ),
            "descendant_output_tokens": sum(
                child.own_output_tokens + child.descendant_output_tokens
                for child in child_summaries
            ),
            "descendant_total_tokens": descendant_total_tokens,
            "descendant_model_cost_usd": descendant_model_cost_usd,
            "budget_status": budget_status,
            "warnings": warnings,
        }
    )


def _lineage_child_run_ids(events: list[AgentStreamEvent]) -> list[str]:
    child_run_ids: list[str] = []
    for event in events:
        if event.type not in LINEAGE_CHILD_EVENT_TYPES or not isinstance(event.payload, dict):
            continue
        child_run_id = event.payload.get("child_run_id")
        if isinstance(child_run_id, str) and child_run_id:
            child_run_ids.append(child_run_id)
    return child_run_ids


def _worst_budget_status(
    summaries: list[AgentRunUsageSummary],
) -> AgentRunBudgetStatus:
    if any(summary.budget_status == "exceeded" for summary in summaries):
        return "exceeded"
    if any(summary.budget_status == "warning" for summary in summaries):
        return "warning"
    return "ok"

File: backend/packages/snapshots/src/usage.ts
import type { AgentStore } from "@aithru-agent/persistence";
import { listChildRuns } from "./tree.js";

type BudgetStatus = "ok" | "warning" | "exceeded";

interface UsageCounters {
  requests: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
}

export interface RunUsageSummary {
  run_id: string;
  own_requests: number;
  own_input_tokens: number;
  own_output_tokens: number;
  own_total_tokens: number;
  descendant_requests: number;
  descendant_input_tokens: number;
  descendant_output_tokens: number;
  descendant_total_tokens: number;
  external_requests: number;
  external_total_tokens: number;
  own_model_cost_usd: number;
  descendant_model_cost_usd: number;
  external_model_cost_usd: number;
  budget_policy: null;
  model_cost_policy: null;
  budget_status: BudgetStatus;
  warnings: string[];
  total_requests: number;
  total_tokens: number;
  total_model_cost_usd: number;
}

export interface RunTreeUsageProjection {
  root_run_id: string;
  runs: RunUsageSummary[];
  total_requests: number;
  total_tokens: number;
  total_model_cost_usd: number;
  budget_status: BudgetStatus;
  warnings: string[];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value != null && typeof value === "object" && !Array.isArray(value);
}

function nonNegativeInt(value: unknown): number {
  if (typeof value !== "number" || !Number.isFinite(value) || value < 0) return 0;
  return Math.trunc(value);
}

function addUsage(a: UsageCounters, b: UsageCounters): UsageCounters {
  return {
    requests: a.requests + b.requests,
    input_tokens: a.input_tokens + b.input_tokens,
    output_tokens: a.output_tokens + b.output_tokens,
    total_tokens: a.total_tokens + b.total_tokens,
  };
}

function usagePayloadCounters(payload: unknown): UsageCounters {
  if (!isRecord(payload)) {
    return { requests: 0, input_tokens: 0, output_tokens: 0, total_tokens: 0 };
  }
  const inputTokens = nonNegativeInt(payload.input_tokens);
  const outputTokens = nonNegativeInt(payload.output_tokens);
  const hasTotalTokens = Object.hasOwn(payload, "total_tokens");
  return {
    requests: nonNegativeInt(payload.requests),
    input_tokens: inputTokens,
    output_tokens: outputTokens,
    total_tokens: hasTotalTokens
      ? nonNegativeInt(payload.total_tokens)
      : inputTokens + outputTokens,
  };
}

function ownUsage(store: AgentStore, runId: string): UsageCounters {
  return store
    .listEvents(runId)
    .filter((event) => event.type === "model.usage")
    .map((event) => usagePayloadCounters(event.payload))
    .reduce(addUsage, { requests: 0, input_tokens: 0, output_tokens: 0, total_tokens: 0 });
}

function ownSummary(runId: string, counters: UsageCounters): RunUsageSummary {
  return {
    run_id: runId,
    own_requests: counters.requests,
    own_input_tokens: counters.input_tokens,
    own_output_tokens: counters.output_tokens,
    own_total_tokens: counters.total_tokens,
    descendant_requests: 0,
    descendant_input_tokens: 0,
    descendant_output_tokens: 0,
    descendant_total_tokens: 0,
    external_requests: 0,
    external_total_tokens: 0,
    own_model_cost_usd: 0,
    descendant_model_cost_usd: 0,
    external_model_cost_usd: 0,
    budget_policy: null,
    model_cost_policy: null,
    budget_status: "ok",
    warnings: [],
    total_requests: counters.requests,
    total_tokens: counters.total_tokens,
    total_model_cost_usd: 0,
  };
}

function sum<T>(items: T[], value: (item: T) => number): number {
  return items.reduce((total, item) => total + value(item), 0);
}

export function buildRunUsageSummary(store: AgentStore, runId: string): RunUsageSummary | undefined {
  if (!store.getRun(runId)) return undefined;
  return ownSummary(runId, ownUsage(store, runId));
}

export function buildRunTreeUsage(store: AgentStore, rootRunId: string): RunTreeUsageProjection | undefined {
  if (!store.getRun(rootRunId)) return undefined;

  const runs: RunUsageSummary[] = [];
  const visiting = new Set<string>();
  const visited = new Set<string>();

  function visit(runId: string): RunUsageSummary | undefined {
    if (visiting.has(runId) || visited.has(runId)) return undefined;
    const summary = buildRunUsageSummary(store, runId);
    if (!summary) return undefined;

    visiting.add(runId);
    visited.add(runId);
    const index = runs.length;
    runs.push(summary);
    const childSummaries = listChildRuns(store, runId)
      .map((child) => visit(child.id))
      .filter((child): child is RunUsageSummary => child != null);
    visiting.delete(runId);

    const descendantRequests = sum(childSummaries, (child) => child.total_requests);
    const descendantInputTokens = sum(
      childSummaries,
      (child) => child.own_input_tokens + child.descendant_input_tokens,
    );
    const descendantOutputTokens = sum(
      childSummaries,
      (child) => child.own_output_tokens + child.descendant_output_tokens,
    );
    const descendantTotalTokens = sum(childSummaries, (child) => child.total_tokens);
    const descendantModelCostUsd = sum(childSummaries, (child) => child.total_model_cost_usd);
    const withDescendants: RunUsageSummary = {
      ...summary,
      descendant_requests: descendantRequests,
      descendant_input_tokens: descendantInputTokens,
      descendant_output_tokens: descendantOutputTokens,
      descendant_total_tokens: descendantTotalTokens,
      descendant_model_cost_usd: descendantModelCostUsd,
      total_requests: summary.own_requests + descendantRequests + summary.external_requests,
      total_tokens: summary.own_total_tokens + descendantTotalTokens + summary.external_total_tokens,
      total_model_cost_usd:
        summary.own_model_cost_usd + descendantModelCostUsd + summary.external_model_cost_usd,
    };
    runs[index] = withDescendants;
    return withDescendants;
  }

  const root = visit(rootRunId);
  if (!root) return undefined;
  return {
    root_run_id: rootRunId,
    runs,
    total_requests: root.total_requests,
    total_tokens: root.total_tokens,
    total_model_cost_usd: root.total_model_cost_usd,
    budget_status: "ok",
    warnings: [],
  };
}
